src/re/multi_grid/kernel.py

```diff
@@ -0,0 +1,453 @@
+
+    def optimize(
+        self,
+        *,
+        rtol=1e-5,
+        atol=1e-10,
+        buffer_size=10000,
+        use_distances=True,
+        distance_norm=partial(jnp.linalg.norm, axis=0),
+    ):
+        """Optimize the kernel for the given grid.
+
+        This links kernel matrices by similarity and subsequently looks them up via
+        tables.
+        """
+
+        def get_distance_matrices(index, level):
+            if level is None:
+                raise NotImplementedError
+            (out, olvl), ids = self.get_output_input_indices(index, level)
+            out = out.reshape(index.shape + (-1,))
+            out = self.grid.at(olvl).index2coord(out)
+            assert index.ndim == out.ndim - 1
+            ids = tuple(self.grid.at(ii[1]).index2coord(ii[0]) for ii in ids)
+            ids = jnp.concatenate(ids, axis=-1)
+            assert index.ndim == ids.ndim - 1
+            return (distance_norm(out[..., jnp.newaxis] - ids[..., jnp.newaxis, :]),)
+
+        gridf = self.grid if isinstance(self.grid, FlatGrid) else FlatGrid(self.grid)
+        uindices = []
+        invindices = []
+        indexmaps = []
+        for lvl in range(self.grid.depth):
+            grid_at = self.grid.at(lvl)
+            gridf_at = gridf.at(lvl)
+
+            def get_matrices(idx):
+                f = get_distance_matrices if use_distances else self.get_matrices
+                ker = f(gridf_at.flatindex2index(jnp.atleast_1d(idx)), lvl)
+                ker = jnp.concatenate(tuple(kk.ravel() for kk in ker))
+                return ker
+
+            @jit
+            def scanned_amend_unique(carry, idx, shift):
+                (u, inv) = carry
+                k = get_matrices(idx)
+                u, invid = amend_unique_(u, k, axis=0, atol=atol, rtol=rtol)
+                inv = inv.at[idx - shift].set(invid)
+                return (u, inv), invid
+
+            indices = grid_at.refined_indices()
+            indices = gridf_at.index2flatindex(indices)[0].ravel()
+            shift = np.min(indices)
+            size = np.max(indices) - shift
+            inv = np.full(size, buffer_size + 1)
+
+            shp = eval_shape(get_matrices, indices[0]).shape
+            unique = jnp.full((buffer_size,) + shp, jnp.nan)
+
+            (unique, inv), invid = scan(
+                partial(scanned_amend_unique, shift=shift), (unique, inv), indices
+            )
+            _, idx = np.unique(invid, return_index=True)
+            n = idx.size
+            if n >= unique.shape[0] or not np.all(np.isnan(unique[n:])):
+                raise ValueError("`mat_buffer_size` too small")
+            uids = indices[idx]
+            uids = gridf_at.flatindex2index(uids[np.newaxis, :])
+            uindices.append(uids)
+            invindices.append(inv)
+            indexmaps.append(_IdxMap(shift, gridf_at.index2flatindex))
+        return OptimizedKernel(self, uindices, invindices, indexmaps)
+
+
+class OptimizedKernel(Kernel):
```

Comment:
Any chance we find a more descriptive name than OptimizedKernel? I'd like to reserve optimize for things that relate to parameter optimizations. Any thoughts on CompressedKernel, BinnedKernel?

Reply:
:+1: I think `CompressedKernel` is a very descriptive and accessible name.